File: 2_pthread/cases.py
import os
import subprocess
import re

inputs = [
    './datasets/sample.in',
    './datasets/1k.in',
    './datasets/20k.in',
    './datasets/20k2k.in',
    './datasets/input1.txt',
    './datasets/input2.txt',
    './datasets/input3.txt'
]

threads = [
    '1', '2', '4', '8', '16'
]

out = {k: {t: {'score': None, 'time': None, 'wrong': False}
           for t in threads + ['serial']} for k in inputs}


def run(cmd):
    proc = subprocess.Popen(cmd,
                            stdout=subprocess.PIPE,
                            stderr=subprocess.PIPE,
                            cwd=os.getcwd()
                            )
    stdout, stderr = proc.communicate()

    return proc.returncode, stdout, stderr


def main():
    print('compile')
    os.system(
        'g++ -std=c++11 -pthread main.cpp pthreads_smith_waterman.cpp -o pthreads_smith_waterman')

    for inp in inputs:
        _, stdout, stderr = run(['./serial/serial_smith_waterman', inp])

        on_std_out(stdout.decode('utf-8'),
                   './serial/serial_smith_waterman', inp, 'serial')
        on_std_err(stderr.decode('utf-8'),
                   './serial/serial_smith_waterman', inp, 'serial')

        for thread in threads:
            print(thread, inp)
            _, stdout, stderr = run(['./pthreads_smith_waterman', inp, thread])
            on_std_out(stdout.decode('utf-8'),
                       './pthreads_smith_waterman', inp, thread)
            on_std_err(stderr.decode('utf-8'),
                       './pthreads_smith_waterman', inp, thread)
    printWhenDone()


def on_std_out(stdout, cmd, inp, num):
    stdout = re.sub(r'\s+', '', stdout)
    if re.match(r'\d+', stdout):
        score = re.findall(r'\d+', stdout)[0]
        out[inp][num]['score'] = score
    else:
        out[inp][num]['score'] = ''
        print(cmd, inp, num, stdout)


def on_std_err(stderr, cmd, inp, num):
    stderr = re.sub(r'\s+', '', stderr)
    if re.match(r'^Time:([\d.e-]+)s$', stderr):
        time = re.findall(r'^Time:([\d.e-]+)s$', stderr)[0]
        out[inp][num]['time'] = time
    else:
        out[inp][num]['time'] = ''
        print(cmd, inp, num, stderr)


def printWhenDone():
    wrong = []
    for inp in inputs:
        for thread in threads:
            if out[inp][thread]['score'] != out[inp]['serial']['score']:
                print(inp, thread, out[inp][thread]
                      ['score'], out[inp]['serial']['score'])
                out[inp][thread]['wrong'] = True
                wrong.append((inp, thread))
    f = chr(10).join(map(
        lambda thread: f"| n = {thread} | {' | '.join(map(lambda i: out[i][thread]['time'], inputs))} | ", threads))
    md = f"\
    # Time \n\
           \n\
    |        | {' | '.join(map(lambda i: '`' + i + '`', inputs))} | \n\
    | ------ | {' | '.join(map(lambda i: ' ----- ', inputs))} | \n\
    | serial | {' | '.join(map(lambda i: out[i]['serial']['time'], inputs))} | \n\
    {f}    "

    print(md)
    with open('time.md', 'w') as f:
        f.writelines(md)


if __name__ == '__main__':
    main()

File: 2_pthread/cases.js
const { execSync, spawnSync } = require('child_process')
const fs = require('fs')

const inputs = [
  'sample.in',
  '1k.in',
  '20k.in',
  '20k2k.in',
  'input1.txt',
  'input2.txt',
  'input3.txt'
].map(input => './datasets/' + input)

const threads = [
  '1', '2', '4', '8', '16'
]

const out = {}

const main = () => {
  inputs.forEach(input => {
    const empty = { score: null, time: null }
    out[input] = { serial: { ...empty } }
    threads.forEach(thread => {
      out[input][thread] = { ...empty }
    })
  })

  console.log('compile')
  execSync('g++ -std=c++11 -pthread main.cpp pthreads_smith_waterman_skeleton.cpp -o pthreads_smith_waterman')

  inputs.forEach(input => {
    console.log('serial', input)
    const { stdout, stderr } = spawnSync('./serial/serial_smith_waterman', [input],)
    onStdOut(stdout.toString(), './serial/serial_smith_waterman', input, 'serial')
    onStdErr(stderr.toString(), './serial/serial_smith_waterman', input, 'serial')
    threads.forEach(thread => {
      console.log(thread, input)
      const { stdout, stderr } = spawnSync('./pthreads_smith_waterman', [input, thread],)
      onStdOut(stdout.toString(), './pthreads_smith_waterman', input, thread)
      onStdErr(stderr.toString(), './pthreads_smith_waterman', input, thread)
    })
  })
  printWhenDone()
}

const onStdOut = (str, cmd, input, num) => {
  const pstr = str.replace(/\s+/g, '')
  if (/^(\d+)$/.test(pstr)) {
    const score = Number.parseInt(pstr.match(/\d+/)[0])
    out[input][num].score = score
  } else {
    out[input][num].score = ''
    console.log(cmd, input, num, str)
  }
}

const onStdErr = (str, cmd, input, num) => {
  const pstr = str.replace(/\s+/g, '')
  if (/^Time:([\d.e-]+)s$/.test(pstr)) {
    const time = Number.parseFloat(pstr.match(/^Time:([\d.e-]+)s$/)[1])
    out[input][num].time = time
  } else {
    out[input][num].time = ''
    console.log(cmd, input, num, str)
  }
}

const deepExists = (object, keyword) => {
  if (object === null) {
    return keyword === null
  } else if (typeof object === 'object') {
    return Object.keys(object)
      .reduce(
        (flag, key) => {
          if (flag) {
            return flag
          } else {
            return deepExists(object[key], keyword)
          }
        }, false
      )
  } else {
    return `${object}`.toLowerCase().includes(keyword)
  }
}

const printWhenDone = () => {
  if (deepExists(out, null)) return
  // console.log(out)
  const wrong = []
  inputs.forEach(input => {
    threads.forEach(thread => {
      if (out[input][thread].score !== out[input].serial.score) {
        console.log(input, thread, out[input][thread].score, out[input].serial.score)
        out[input][thread].wrong = true
        wrong.push({ input, thread })
      }
    })
  })
  const md = `
  # Time
  
  |        | ${inputs.map(i => '`' + i + '`').join(' | ')} |
  | ------ | ${inputs.map(i => ' ----- ').join(' | ')} |
  | serial | ${inputs.map(input => out[input].serial.time).join(' | ')} |
  ${threads.map(thread => `\
  | n=${thread} | ${inputs.map(input => out[input][thread].time).join(' | ')} |\
  `).join('\n')}
  `
  console.log(md)
  fs.writeFile('time.md', md, err => {
    if (err === null) {
      console.log('successully write to time.md')
    }
  })
  if (wrong.length > 0) {
    console.log(wrong)
  }
}

process.on('exit', printWhenDone.bind(null, { out }))

main()
